File: Permutation in String.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        map<char,int> m;
        vector<int> x;
        for(int i=0;i<s1.length();i++)
            m[p[i]]++;
        int l=0,r;
        int cnt=s1.length();
        for(int r=0;r<s2.length();r++)
        {
            if(m.find(s2[r])!=m.end() && m[s2[r]])
            {
               m[s2[r]]--;
                cnt--;
                if(cnt==0)
                    return true;
            }
            else{
                while(l<r)
                {
                    if(m.find(s2[l])!=m.end() )
                    {
                        m[s2[l]]++;
                        cnt++;
                    }
                    l++;
                    if(m.find(s2[r])!=m.end() && m[s2[r]])
                    {
                        r--;
                        break;
                    }
                }
                if(m.find(s2[l])==m.end())
                    l=r+1;
            }
           
            
        }
        return false;
        
    }
    
};

File: Golang Solutions/Permutation in String.go
func checkInclusion(s1 string, s2 string) bool {
    m:=make(map[byte]int)
    for i:=0;i<len(s1);i++{
        _,ok:=m[s1[i]]
        if !ok{
            m[s1[i]]=1
        }else{
            m[s1[i]]++
        }
    }
    l,r:=0,0
    cnt:=len(s1)
    for r=0;r<len(s2);r++{
        a,ok:= m[s2[r]]
        if ok && a>0{
            m[s2[r]]--
            cnt--
            if cnt==0{
                return true
            }
                    
        }else{
            for ;l<r;{
                _,ok1:=m[s2[l]]
                if ok1 {
                    m[s2[l]]++
                    cnt++
                }
                l++
                b,ok2:=m[s2[r]]
                if ok2 && b>0{
                    r--;
                    break;
                }
            }
            _,ok3:=m[s2[l]]
            if !ok3{
                l=r+1
            }    
        }
    }
    return false
    
}